### ISSUE TITLE
[BUG] MOD_SET_DX times out on MTL platform

### ISSUE THREAD

**Issue by keqiaozhang** [thesofproject/linux#4246]
**Describe the bug**
This issue happened on MTL when testing suspend/resume related cases. From the demsg, it failed with MOD_SET_DX timeout.
```
[  142.743284] kernel: snd_sof:sof_ipc4_log_header: sof-audio-pci-intel-mtl 0000:00:1f.3: ipc tx      : 0x44000000|0x3060004c: MOD_LARGE_CONFIG_SET [data size: 76]
[  142.743892] kernel: snd_sof:sof_ipc4_log_header: sof-audio-pci-intel-mtl 0000:00:1f.3: ipc tx reply: 0x64000000|0x3060004c: MOD_LARGE_CONFIG_SET
[  142.743966] kernel: snd_sof:sof_ipc4_log_header: sof-audio-pci-intel-mtl 0000:00:1f.3: ipc tx done : 0x44000000|0x3060004c: MOD_LARGE_CONFIG_SET [data size: 76]
[  142.744060] kernel: snd_sof:sof_ipc4_log_header: sof-audio-pci-intel-mtl 0000:00:1f.3: ipc tx      : 0x47000000|0x0: MOD_SET_DX [data size: 8]
[  143.247354] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: ipc timed out for 0x47000000|0x0
[  143.247443] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: preventing DSP entering D3 state to preserve context
[  143.247454] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: ------------[ IPC dump start ]------------
[  143.247552] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: Host IPC initiator: 0x47000000|0x0|0x0, target: 0x64000000|0x3060004c|0x0, ctl: 0x3
[  143.247634] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: ------------[ IPC dump end ]------------
[  143.247689] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: ------------[ DSP dump start ]------------
[  143.247744] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: IPC timeout
[  143.247785] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: fw_state: SOF_FW_BOOT_COMPLETE (7)
[  143.247864] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: ROM status: 0x5, ROM error: 0x0
[  143.247932] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: ROM debug status: 0x50000005, ROM debug error: 0x0
[  143.247997] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: ROM feature bit enabled
[  143.248043] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: ------------[ DSP dump end ]------------
[  143.248101] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: error: -110 failed to prepare DSP for device removal
```
Before the reproduction rate is about 50%, but after https://github.com/thesofproject/sof/commit/0664718a5896abe1e50191218cb80abb7171ad2d, this issue can be reproduced on MTL 100% at the first round of suspend/resume.

**To Reproduce**
rtcwake -m mem -s 5 

**Reproduction Rate**
<del>100% after https://github.com/thesofproject/sof/commit/0664718a5896abe1e50191218cb80abb7171ad2d, before the repro rate is about 50%.</del>
[UPDATE]: the reproduction rate has been extremely low since Dec 23 on tip sof/main.

**Expected behavior**

**Environment**
1) Branch name and commit hash of the 2 repositories: sof (firmware/topology) and linux (kernel driver).
    * Kernel: tip sof-dev/eebcfac2c69e
    * SOF: main/99a65fc7d72f
2) Name of the platform(s) on which the bug is observed.
    * Platform: MTL

[dmesg.txt](https://github.com/thesofproject/sof/files/10233416/dmesg.txt)

**Comment by kfrydryx** [comment]
DX functionalities on MTL are in progress, issues are expected at the moment. Let's keep it on hold and revisit a bit later.

**Comment by keqiaozhang** [comment]
This issue also can be reproduced on TGL-IPC4:
https://sof-ci.sh.intel.com/#/result/planresultdetail/19395?model=TGLU_RVP_SDW_IPC4ZPH&testcase=check-suspend-resume-with-playback-5

**Comment by miRoox** [comment]
The reproduction rate become pretty low, and now we usually see https://github.com/thesofproject/linux/issues/4077 instead.

**Comment by keqiaozhang** [comment]
This issue is not reproducible in CI, closing this bug.

**Comment by miRoox** [comment]
It happened again today, seems the reproduction rate is very low.

**Comment by mengdonglin** [comment]
Deprioritize it to P2 due to the very low reproduce rate atm.

**Comment by keqiaozhang** [comment]
The error changed from `MOD_SET_DX` to `MOD_SET_D0IX`. This error change seems to have occurred after recent upstream mergers.
```
[ 1260.565574] kernel: snd_sof:sof_ipc4_log_header: sof-audio-pci-intel-mtl 0000:00:1f.3: ipc tx      : 0x44000000|0x3060004c: MOD_LARGE_CONFIG_SET [data size: 76]
[ 1260.566134] kernel: snd_sof:sof_ipc4_log_header: sof-audio-pci-intel-mtl 0000:00:1f.3: ipc tx reply: 0x64000000|0x3060004c: MOD_LARGE_CONFIG_SET
[ 1260.566169] kernel: snd_sof:sof_ipc4_log_header: sof-audio-pci-intel-mtl 0000:00:1f.3: ipc tx done : 0x44000000|0x3060004c: MOD_LARGE_CONFIG_SET [data size: 76]
[ 1260.566266] kernel: snd_sof:sof_ipc4_log_header: sof-audio-pci-intel-mtl 0000:00:1f.3: ipc tx      : 0x48000000|0x12: MOD_SET_D0IX
[ 1261.072884] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: ipc timed out for 0x48000000|0x12
[ 1261.072894] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: preventing DSP entering D3 state to preserve context
[ 1261.072897] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: ------------[ IPC dump start ]------------
[ 1261.072906] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: Host IPC initiator: 0xc8000000|0x12|0x0, target: 0x64000000|0x3060004c|0x0, ctl: 0x3
[ 1261.072910] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: ------------[ IPC dump end ]------------
[ 1261.072912] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: ------------[ DSP dump start ]------------
[ 1261.072913] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: IPC timeout
[ 1261.072915] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: fw_state: SOF_FW_BOOT_COMPLETE (7)
[ 1261.072922] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: ROM status: 0x5, ROM error: 0x0
[ 1261.072924] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: ROM debug status: 0xa, ROM debug error: 0x0
[ 1261.072927] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: ROM feature bit enabled
[ 1261.072928] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: ------------[ DSP dump end ]------------
[ 1261.072934] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: error: PM_GATE ipc error -110
[ 1261.073020] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: failed to set requested target DSP state 0 substate 1
[ 1261.073032] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: error: setting dsp state 0 substate 1
[ 1261.073042] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: error: failed to power down DSP during suspend -110
[ 1261.073052] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: PM: pci_pm_suspend(): snd_sof_suspend+0x0/0x20 [snd_sof] returns -110
[ 1261.073157] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: PM: dpm_run_callback(): pci_pm_suspend+0x0/0x160 returns -110
[ 1261.073191] kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: PM: failed to suspend async: error -110
[ 1261.073448] kernel: PM: Some devices failed to suspend, or early wake event detected
```

**Comment by kv2019i** [comment]
V2.5 branching this week and nobody assigned, so moving to v2.6.

**Comment by fredoh9** [comment]
https://github.com/thesofproject/linux/pull/4245 fixed the problem. already verified with MTL systems.

**Comment by ujfalusi** [comment]
if #4245 fixed this then it is only by luck and the fact that we have tracing enabled in CI, if you don't have BIT(0) set in sof_debug module parameter (normal use) then the MOD_LARGE_CONFIG_SET would not be sent prior to SET_DX since that MOD_LARGE_CONFIG_SET is the disable mtrace message.

To sum up: #4245 did not fixed anything related to this bug and we have now a works by luck solution to please CI.

**Comment by ranj063** [comment]
> To sum up: #4245 did not fixed anything related to this bug and we have now a works by luck solution to please CI.

@ujfalusi that assessment is incorrect. The PR does not work by luck. You cannot send the MOD_LARGE_CONFIG_SET IPC during D0i3 without bring the DSP back to D0i0. It is isnt a made up situation either. In the case of the CI tests, it is due to disabling trace but you could very well send kcontrol data when the deep buffer PCM is active and the DSP is in D0i3 and you'll hit ther same problem.

**Comment by ujfalusi** [comment]
@ranj063, I don't question that, we need to bring the DSP to D0I0 for the LARGE_CONFIG message, yes.
What I'm saying is that the bug is stating the we have an issue with the SET_DX, which is the last message, sent after the LARGE_CONFIG (which is the disable mtrace message), if we disable the tracing (bit(0) in sof_debug) then there will be no prior message to SET_DX and the timeout might be there still.

It can be easily tested with clearing the bit(0) in sof_debug, reload the drivers and repeat the test.

I might be well be wrong, but this is how it looks to me.

**Comment by fredoh9** [comment]
I was able to reproduce the problem with sof_debug=0x80. It wasn't easy to reproduce, had to do multiple runs.

**Comment by ranj063** [comment]
@ujfalusi so we have 2 separate problems here. An IPC timeout when trace is enabled which happens because the LARGE_CONFIG_SET is sent during D0i3 and is fixed by #4245. Additionally, it looks like the SOF firmware ends up in a PM ref count imbalance because of multiple MOD_D0IX IPCs sent to set the state to D0i3 when the DSP is already in D0i3. #4257 should handle that case as well.
